common/rate-limiter/src/rate_limit.rs

```diff
@@ -0,0 +1,498 @@
+        default_fill_rate: usize,
+    ) -> Self {
+        // Ensure that we can actually use the rate limiter
+        assert!(new_bucket_start_percentage <= 100);
+        assert!(default_bucket_size > 0);
+        assert!(default_fill_rate > 0);
+
+        Self {
+            label,
+            buckets: RwLock::new(HashMap::new()),
+            new_bucket_start_percentage,
+            default_bucket_size,
+            default_fill_rate,
+            open: false,
+        }
+    }
+
+    pub fn test(default_bucket_size: usize, default_fill_rate: usize) -> Self {
+        Self::new("test", 100, default_bucket_size, default_fill_rate)
+    }
+
+    /// Used for testing and to not have a rate limiter
+    pub fn open(label: &'static str) -> Self {
+        Self {
+            label,
+            buckets: RwLock::new(HashMap::new()),
+            new_bucket_start_percentage: 100,
+            default_bucket_size: std::usize::MAX,
+            default_fill_rate: std::usize::MAX,
+            open: true,
+        }
+    }
+
+    /// Retrieve bucket, or create a new one
+    pub fn bucket(&self, key: Key) -> SharedBucket {
+        if !self.open {
+            self.bucket_inner(key, |label, initial, size, rate| {
+                Arc::new(Mutex::new(Bucket::new(label, initial, size, rate)))
+            })
+        } else {
+            self.bucket_inner(key, |label, _, _, _| {
+                Arc::new(Mutex::new(Bucket::open(label)))
+            })
+        }
+    }
+
+    fn bucket_inner<F: FnOnce(String, usize, usize, usize) -> SharedBucket>(
+        &self,
+        key: Key,
+        bucket_create: F,
+    ) -> SharedBucket {
+        // Attempt to do a weaker read lock first, followed by a write lock if it's missing
+        // For the common (read) case, there should be higher throughput
+        // Note: This read must happen in a separate block, to ensure the read unlock for the write
+        let maybe_bucket = { self.buckets.read().get(&key).cloned() };
+        if let Some(bucket) = maybe_bucket {
+            bucket
+        } else {
+            let size = self.default_bucket_size;
+            let rate = self.default_fill_rate;
+
+            // Write in a bucket, but make sure again that it isn't there first
+            self.buckets
+                .write()
+                .entry(key.clone())
+                .or_insert_with(|| {
+                    let label = Self::stringify_key(self.label, key);
+                    bucket_create(
+                        label,
+                        size.saturating_mul(self.new_bucket_start_percentage as usize) / 100,
+                        size,
+                        rate,
+                    )
+                })
+                .clone()
+        }
+    }
+
+    /// Garbage collects a single key, if we know what it is
+    pub fn try_garbage_collect_key(&self, key: &Key) -> bool {
+        let mut write = self.buckets.write();
+        let arc = write.get(key);
```

Comment:
might be better to make a single map that does all this?
but also arc => maybe key
```
buckets.get(key).map(|bucket| if ... { write.remove(key); true } else { false }).unwrap_or(false);

Reply:
Yeah, these terms aren't the greatest let me make a followup to fix some of these.

Reply:
I remember now, I tried to use a map statement, and it doesn't work because dealing with making a mutable borrow within another mutable borrow.